### ISSUE TITLE
Support citation of full Relaton references

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/metanorma#52]
In Metanorma XML, there are actually two types of references:
* Full references, which use the Relaton data model
* Simple references, the ones manually inputted.
 
`<formattedref>` used for simple references is actually a “hack” — it is only used when the full bibliographic data isn’t available.
 
This is best demonstrated with examples.
 
In Metanorma AsciiDoc files, if you see:
```adoc
* [[[iso9001,ISO 9001]]], _foo_
```
 
The `{publisher} {id}` is a special format that only supports certain publishers (SDOs like ISO, IEC, ITU, NIST,…). Metanorma knows to fetch `ISO 9001` from iso.org, and will embed the “full reference” that includes all data elements necessary for rendering the citation.
 
On the other hand, if you see:
```adoc
* [[[BZIP,BZIP2 Compression Algorithm]], _Entropy Measures and Unconditional Security in Cryptography_, C. Cachin, PhD Thesis, Reprint as vol.1 of ETH Series in Information Security and Cryptography, ISBN 3-89649-185-7, Hartung-Gorre Verlag, Konstanz, ETH Zurich, 1997.
```
 
This is purely manual input. For the latter type of citation, we can only render it using `<formattedref>` because we cannot fit the input information into a citation data model.
 
The reason we allow this is due to limitations of the original AsciiDoc format.
 
This issue is to allow Metanorma users to create Relaton citations for any type of reference.

Prerequisites:

* Our `asciidoctor-bibliography` gem (https://github.com/riboseinc/asciidoctor-bibliography) allows us to read BibTeX files.
* Our `relaton` gem supports reading Relaton YAML files.

1. External Relaton YAML file. In Metanorma, we should be able to specify a "Relaton" (YAML/XML) file or directory that contains bibliographic information in the Relaton model.

e.g.
```
:bib-directory-relaton: ~/relaton/
```

or

```
:bib-file-relaton: ~/relaton.yaml
```

2. External BibTeX file (and EndNote, RefMan, RefWorks...)

Our `asciidoctor-bibliography` gem (https://github.com/riboseinc/asciidoctor-bibliography) allows us to read BibTeX files. This flow is extended from Relaton, where:

```
BibTeX -> Relaton model -> used in Metanorma
```
 
@CAMOBAP795 can you help handle these two tasks described above?

cc: @opoudjis

**Comment by ronaldtse** [comment]
e.g. For the second case, suppose I have this in a BibTeX file:

```
@article{dworkin2016recommendation,
  title={Recommendation for block cipher modes of operation: methods for format-preserving encryption},
  author={Dworkin, Morris},
  journal={NIST Special Publication},
  volume={800},
  pages={38G},
  year={2016}
}
```

Then I can do this in Metanorma:

```adoc
The criteria is listed in <<dworkin2016recommendation>> ...
```

**Comment by opoudjis** [comment]
This is distinct from what I will end up having to do, which is render those citations, with an idiosyncratic stylesheet no doubt for each goddamn standard.

**Comment by CAMOBAP795** [comment]
In progress

**Comment by ronaldtse** [comment]
> This is distinct from what I will end up having to do, which is render those citations, with an idiosyncratic stylesheet no doubt for each goddamn standard.

@opoudjis we need a bit more discussion on this. There is often a designated citation style of the publication (e.g. ISO 690 for ISO documents) that should be applied. I'm not sure why we need an idiosyncratic stylesheet?

**Comment by opoudjis** [comment]
The stylesheet is specific to the standard of the containing document, not the standard of the cited reference.

**Comment by ronaldtse** [comment]
Right, and by "stylesheet" here you mean "citation rendering rules". Indeed.

**Comment by CAMOBAP795** [comment]
@opoudjis just to be on the same page, because you took https://github.com/metanorma/relaton/issues/33, should I leave this one for you?

**Comment by ronaldtse** [comment]
Further clarification -- the citation rendering should utilize CSL, which is a combination of:
* https://github.com/inukshuk/citeproc
* https://github.com/inukshuk/csl-ruby

We will need to create a ISO 690 CSL schema because the current one cannot produce ISO 690 style (esp. citation of an ISO Standard).

**Comment by opoudjis** [comment]
@ronaldtse is once again racing much too far ahead in his last comment, and there are a lot more connective steps that need to be put in place.

I desperately want to offload some work, but work cannot be offloaded until it is ready. And there is some unacceptable handwaving in this ticket.

This ticket falls under the stuff I tentatively assigned to Abu Nashir in https://github.com/metanorma/relaton-cli/wiki/Program-of-work,-May-2019. Until Andrej is finished implementing the refactoring of the relaton classes (which he has just started), this task is not to be begun. The first step of this process is Relaton YAML > Relaton class constructor. The Relaton Definition List implementation will feed into that.

There are a bunch of other steps that need to be specified a lot better than they have in this ticket. And CSL support is not a throway afterthought, it is itself a major undertaking. One, I note with some annoyance, whose integration with Metanorma has not even been discussed.

@ronaldtse, kindly review the Wiki page given above as requested, and add any additional requests like CSL support at the end of it. With thought as to how any CSL output is going to integrate with Metanorma, because right now, I'm doing it as the render module in NIST, and I would need to be convinced that any CSL output is fit for purpose with Metanorma.

This ticket is henceforth restricted to the implementation of Relaton YAML, and any discussion of CSL must take place in a separate ticket. Tickets are not meant to be epics.

**Comment by CAMOBAP** [comment]
@ronaldtse @opoudjis I kind of lost state of this, is anything still need to be done in scope of it?

**Comment by ronaldtse** [comment]
This is a generic task to develop a syntax to enter structured references in a Metanorma bibliography, and related to a ISO 690/citation generation gem.

**Comment by andrew2net** [comment]
> Until Andrej is finished implementing the refactoring of the relaton classes (which he has just started), this task is not to be begun. I'd rather I not take it on, and am happy for you to take it on, @CAMOBAP, but not until @andrew2net is finished: this task presupposes his.

@opoudjis it seems I've missed the task. Could you point where is it?

**Comment by opoudjis** [comment]
This ticket is so old, @andrew2net, it was talking about your creation of relaton classes, read in from hashes. You did this a couple of years ago.
